### PR TITLE
Add client credential clients support

### PR DESCRIPTION
**Pull request by MKodde**
This introduces support for the oauth client credentials client entity type in SP dashboard. This entity type consists of a oidc tng entity that has the client credentials grant. This labels it internally as an `oauth20_ccc` entity.

https://www.pivotaltracker.com/story/show/177492575

### CODE REVIEW

File: src/Surfnet/ServiceProviderDashboard/Application/Metadata/OauthClientCredentialsClientJsonGenerator.php [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        return array_filter($metadata);
+    }
+
+    private function generateContactMetadata($contactType, $index, Contact $contact): array
+    {
+        $metadata = [
+            sprintf('contacts:%d:contactType', $index) => $contactType,
+        ];
+
+        if (!empty($contact->getGivenName())) {
```
Comment (Badlapje) [inline src/Surfnet/ServiceProviderDashboard/Application/Metadata/OauthClientCredentialsClientJsonGenerator.php:266]: All these if statements can be abstracted into a method, they always follow the exact same pattern.

File: src/Surfnet/ServiceProviderDashboard/Domain/Entity/ManageEntity.php [hunk 2]
```diff
@@ -87,15 +89,17 @@ class ManageEntity
...
     {
-        $manageProtocol = isset($data['type']) ? $data['type'] : '';
+        $manageProtocol = self::extractManageProtocol($data);
 
         $attributeList = AttributeList::fromApiResponse($data);
         $metaData = MetaData::fromApiResponse($data);
         $oidcClient = null;
         if ($manageProtocol === Protocol::OAUTH20_RS) {
-            $oidcClient = OidcngResourceServerClient::fromApiResponse($data, $manageProtocol);
+            $oidcClient = OidcngResourceServerClient::fromApiResponse($data);
```
Comment (Badlapje) [inline src/Surfnet/ServiceProviderDashboard/Domain/Entity/ManageEntity.php:98]: A switch would be more readable here, as well as faster.

File: src/Surfnet/ServiceProviderDashboard/Domain/Entity/ManageEntity.php [hunk 3]
```diff
@@ -285,7 +298,10 @@ public function merge(ManageEntity $newEntity)
         $this->metaData->merge($newEntity->getMetaData());
         $this->attributes->merge($newEntity->getAttributes());
         $protocol = $this->protocol->getProtocol();
-        if ($protocol === Constants::TYPE_OPENID_CONNECT_TNG || $protocol === Constants::TYPE_OPENID_CONNECT_TNG_RESOURCE_SERVER) {
+        if ($protocol === Constants::TYPE_OPENID_CONNECT_TNG
```
Comment (Badlapje) [inline src/Surfnet/ServiceProviderDashboard/Domain/Entity/ManageEntity.php:301]: This if would be more readable if you made a method for it.

**Review by Badlapje — CHANGES_REQUESTED** [review]
Nice work!  Looks good, save for a couple of smaller points.

File: src/Surfnet/ServiceProviderDashboard/Domain/Entity/ManageEntity.php (hunk 2 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Domain/Entity/ManageEntity.php]: Agreed

File: src/Surfnet/ServiceProviderDashboard/Application/Metadata/OauthClientCredentialsClientJsonGenerator.php (hunk 1 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Application/Metadata/OauthClientCredentialsClientJsonGenerator.php:266]: I agree, but the method would be rather messy in my opinion. Using the accessors on the contacts would require using something like $contact->$contactGetterName(). That kinda sucks.

Comment (Badlapje) [reply src/Surfnet/ServiceProviderDashboard/Application/Metadata/OauthClientCredentialsClientJsonGenerator.php:266]: hm, a matter of preference whether it's messy or not (i find it a beautiful php feature).  If you dislike it though, it makes little sense to refactor.  I can live with how it is atm.

**Review by Badlapje — APPROVED** [review]
Looking good to me, nice work!  :shipit:

**Pull request merged**